neqo-http3/src/send_message.rs

```diff
@@ -0,0 +1,270 @@
+const MAX_DATA_HEADER_SIZE_5_LIMIT: usize = MAX_DATA_HEADER_SIZE_5 + 9; // 1073741823 + 9 (size of the next buffer data frame header)
+
+pub(crate) trait SendMessageEvents: Debug {
+    fn data_writable(&self, stream_id: u64);
+}
+
+/*
+ *  Transaction send states:
+ *    Initialized : Headers are present but still not encoded. A message body may be present as well.
+ *                  The lcient side send message body using the send_body() function that directly
+ *                  writes into a transport stream. The server side sets headers and body when
+ *                  initializing a send message (TODO: make server use send_body as well)
+ *    SendingInitialMessage : sending headers and maybe meassage body. From here we may switch to
+ *                     SendingData or Closed (if the app does not want to send data and
+ *                     has already closed the send stream).
+ *    SendingData : We are sending request data until the app closes the stream.
+ *    Closed
+ */
+
+#[derive(PartialEq, Debug)]
+enum SendMessageState {
+    Uninitialized,
+    Initialized {
+        headers: Vec<Header>,
+        data: Option<Vec<u8>>,
+        fin: bool,
+    },
+    SendingInitialMessage {
+        buf: Vec<u8>,
+        fin: bool,
+    },
+    SendingData,
+    Closed,
+}
+
+#[derive(Debug)]
+pub(crate) struct SendMessage {
+    state: SendMessageState,
+    stream_id: u64,
+    conn_events: Box<dyn SendMessageEvents>,
+}
+
+impl SendMessage {
+    pub fn new(stream_id: u64, conn_events: Box<dyn SendMessageEvents>) -> Self {
+        qinfo!("Create a request stream_id={}", stream_id);
+        Self {
+            state: SendMessageState::Uninitialized,
+            stream_id,
+            conn_events,
+        }
+    }
+
+    pub fn new_with_headers(
+        stream_id: u64,
+        headers: Vec<Header>,
+        conn_events: Box<dyn SendMessageEvents>,
+    ) -> Self {
+        qinfo!("Create a request stream_id={}", stream_id);
+        Self {
+            state: SendMessageState::Initialized {
+                headers,
+                data: None,
+                fin: false,
+            },
+            stream_id,
+            conn_events,
+        }
+    }
+
+    pub fn set_message(&mut self, headers: &[Header], data: Option<&[u8]>) -> Res<()> {
+        if !matches!(self.state, SendMessageState::Uninitialized) {
+            return Err(Error::AlreadyInitialized);
+        }
+
+        self.state = SendMessageState::Initialized {
+            headers: headers.to_vec(),
+            data: if let Some(d) = data {
+                Some(d.to_vec())
+            } else {
+                None
+            },
+            fin: true,
+        };
+        Ok(())
+    }
+
+    pub fn send_body(&mut self, conn: &mut Connection, buf: &[u8]) -> Res<usize> {
+        qinfo!(
+            [self],
+            "send_request_body: state={:?} len={}",
+            self.state,
+            buf.len()
+        );
+        match self.state {
```

Comment:
I don't think that we need to be as careful with framing at this layer.  If we get blocked, that only prevents us from sending other frames on the stream, which boils down to trailers (which we don't send) and control frames (which don't currently exist).

So I would suggest sending everything we have as possible.  Then you only have to track how much data you have left to send in the current frame (i.e., SendMessageState::SendingData now has a usize parameter).  The only probing you would then need to do is to work out whether the header of the frame fits in the available space, but you can do that with the atomic write.

An application that does have other things to send than data (in the future perhaps) might withhold data from the stack to make space, or we can revise this code and cap the size of writes in some way.

That should make this a lot less complex: Encode a DATA frame header, write it atomically then send whatever data from the provided buffer you can using the non-atomic send.

That also means that an application that attempts to send X bytes and only is able to send X-Y bytes must later send at least Y bytes or bad things will happen because we will otherwise be unable to fill the frame.  It's worth documenting that constraint.

Reply:
It is a good point. I will open a separate issue for this.